x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetDataFrameAnalyticsStatsAction.java

```diff
@@ -166,24 +153,70 @@ protected void doExecute(Task task, GetDataFrameAnalyticsStatsAction.Request req
     void gatherStatsForStoppedTasks(List<String> expandedIds, GetDataFrameAnalyticsStatsAction.Response runningTasksResponse,
                                     ActionListener<GetDataFrameAnalyticsStatsAction.Response> listener) {
         List<String> stoppedTasksIds = determineStoppedTasksIds(expandedIds, runningTasksResponse.getResponse().results());
-        List<Stats> stoppedTasksStats = stoppedTasksIds.stream().map(this::buildStatsForStoppedTask).collect(Collectors.toList());
-        List<Stats> allTasksStats = new ArrayList<>(runningTasksResponse.getResponse().results());
-        allTasksStats.addAll(stoppedTasksStats);
-        Collections.sort(allTasksStats, Comparator.comparing(Stats::getId));
-        listener.onResponse(new GetDataFrameAnalyticsStatsAction.Response(new QueryPage<>(
-            allTasksStats, allTasksStats.size(), GetDataFrameAnalyticsAction.Response.RESULTS_FIELD)));
+        if (stoppedTasksIds.isEmpty()) {
+            listener.onResponse(runningTasksResponse);
+            return;
+        }
+
+        AtomicInteger counter = new AtomicInteger(stoppedTasksIds.size());
+        AtomicArray<Stats> stoppedStats = new AtomicArray<>(stoppedTasksIds.size());
+        for (int i = 0; i < stoppedTasksIds.size(); i++) {
+            int slot = i;
+            String analyticsId = stoppedTasksIds.get(i);
+            searchStoredProgress(analyticsId, ActionListener.wrap(
```

Comment:
I think it would be better to have a single query that gathers all the stored progresses for all the stopped analytics processes. The only concern would be query size, but I think we are restricted to 10k already given the ID expansion logic.

Reply:
I thought about it, `size` concerns made me lazy, but you're right. Shouldn't be a problem. Will change.